lint_rules/lint_rules/debug_rules.py

```diff
@@ -0,0 +1,296 @@
+                return True
+
+        return False
+
+    @classmethod
+    def get_explicit_arg_size(cls, arg, dims):
+        """
+        Method to return the size of a subroutine argument whose bounds are
+        explicitly declared.
+        """
+
+        if isinstance(arg, Scalar):
+            size = as_tuple(IntLiteral(1))
+        else:
+            size = ()
+            for dim in dims:
+                if isinstance(dim, RangeIndex):
+                    size += as_tuple(simplify(cls.range_to_sum(dim.lower, dim.upper)))
+                else:
+                    size += as_tuple(dim)
+
+        return size
+
+    @classmethod
+    def get_implicit_arg_size(cls, arg, dims):
+        """
+        Method to return the size of a subroutine argument whose bounds are
+        potentially implicitly declared.
+        """
+
+        size = ()
+        for count, dim in enumerate(dims):
+            if isinstance(dim, RangeIndex):
+                if not dim.upper:
+                    if isinstance(arg.shape[count], RangeIndex):
+                        upper = arg.shape[count].upper
+                    else:
+                        upper = arg.shape[count]
+                else:
+                    upper = dim.upper
+                if not dim.lower:
+                    if isinstance(arg.shape[count], RangeIndex):
+                        lower = arg.shape[count].lower
+                    else:
+                        lower = IntLiteral(1)
+                else:
+                    lower = dim.lower
+
+                size += as_tuple(cls.range_to_sum(lower, upper))
+            else:
+                size += as_tuple(dim)
+
+        return size
+
+    @classmethod
+    def check_subroutine(cls, subroutine, rule_report, config, **kwargs):
+        """
+        Method to check for argument size mismatches across subroutine calls.
+        It requires all :any:`CallStatement` nodes to be enriched, and requires
+        all subroutine arguments *to not be* of type :any:`DeferredTypeSymbol`.
+        Therefore relevant modules should be parsed before parsing the current
+        :any:`Subroutine`.
+        """
+
+        # first resolve associates
+        resolve_associates(subroutine)
+
+        assign_map = {a.lhs: a.rhs for a in FindNodes(Assignment).visit(subroutine.body)}
+        decl_symbols = flatten([decl.symbols for decl in FindNodes(VariableDeclaration).visit(subroutine.spec)])
+        decl_symbols = [sym for sym in decl_symbols if sym.type.initial]
+        assign_map.update({sym: sym.initial for sym in decl_symbols})
+
+        targets = as_tuple(kwargs.get('targets', None))
+        calls = [c for c in FindNodes(CallStatement).visit(subroutine.body) if c.name in targets]
+
+        for call in calls:
+
+            # check if calls are enriched
+            if not call.routine:
+                continue
+
+            arg_map = {carg: rarg for rarg, carg in call.arg_iter()}
+            for arg in arg_map:
```

Comment:
You could even go further and use `for carg, rarg in arg_map.items()` to save the `arg_map[arg]` lookups in the loop body. But mentioning this only for educational purposes.